tools/gulp/tasks/screenshots.ts

```diff
@@ -0,0 +1,190 @@
+ *   For golds mode should be null.
+ */
+function uploadScreenshots(reportKey?: string, mode?: 'test' | 'diff') {
+  let bucket = openScreenshotsCloudStorage();
+
+  let promises: admin.Promise<void>[] = [];
+  let localDir = mode == 'diff' ? `${SCREENSHOT_DIR}/diff` : SCREENSHOT_DIR;
+  readdirSync(localDir).map((file: string) => {
+    let fileName = path.join(localDir, file);
+    let key = file.replace('.screenshot.png', '');
+    let destination = (mode == null || !reportKey) ?
+      `golds/${file}` : `screenshots/${reportKey}/${mode}/${file}`;
+
+    if (!statSync(fileName).isDirectory() && key) {
+      promises.push(bucket.upload(fileName, { destination: destination }));
+    }
+  });
+  return admin.Promise.all(promises);
+}
+
+/** Check whether the directory exists. If not then create one. */
+function _makeDir(dirName: string) {
+  if (!existsSync(dirName)) {
+    mkdirSync(dirName, '744');
+  }
+}
+
+/** Download golds screenshots. */
+function downloadAllGolds(
+    filenames: string[], database: admin.database.Database,
+    reportKey: string): admin.Promise<boolean> {
+  _makeDir(`${SCREENSHOT_DIR}/golds`);
+
+  return admin.Promise.all(filenames.map((filename: string) => {
+    return downloadGold(filename).then(() => diffScreenshot(filename, database, reportKey));
+  })).then((results: boolean[]) => results.every((value: boolean) => value == true));
+}
+
+/** Download one gold screenshot */
+function downloadGold(filename: string): Promise<void> {
+  let bucket = openScreenshotsCloudStorage();
+  return bucket.file(`golds/${filename}`).download({
+    destination: `${SCREENSHOT_DIR}/golds/${filename}`
+  });
+}
+
+function diffScreenshot(filename: string, database: admin.database.Database,
+                             reportKey: string): admin.Promise<boolean> {
+  // TODO(tinayuangao): Run the downloads and diffs in parallel.
+  let goldUrl = `${SCREENSHOT_DIR}/golds/${filename}`;
+  let pullRequestUrl = `${SCREENSHOT_DIR}/${filename}`;
+  let diffUrl = `${SCREENSHOT_DIR}/diff/${filename}`;
+  let filenameKey = filename.replace('.screenshot.png', '');
+
+  if (existsSync(goldUrl) && existsSync(pullRequestUrl)) {
+    return new admin.Promise((resolve: any, reject: any) => {
+      imageDiff({
+        actualImage: pullRequestUrl,
+        expectedImage: goldUrl,
+        diffImage: diffUrl,
+      }, (err: any, imagesAreSame: boolean) => {
+        if (err) {
+          console.log(err);
+          imagesAreSame = false;
+          reject(err);
+        }
+        resolve(imagesAreSame);
+        return updateFileResult(database, reportKey, filenameKey, imagesAreSame);
+      });
+    });
+  } else {
+    return updateFileResult(database, reportKey, filenameKey, false).then(() => false);
+  }
+}
+
+function decode(value: string): string {
+  return value.split('').reverse().join('');
+}
+
+function updateGithubStatus(result: boolean, prNumber: string) {
```

Comment:
This is a very nice function! Please move this into a utility function. e.g in `tools/gulp/util-functions.ts`

Be sure to remove the `new admin.Promise` and just use `new Promise` from ES6